pkg/trie/triedb/node.go

```diff
@@ -168,8 +207,204 @@ func newNodeFromEncoded(nodeHash common.Hash, data []byte, storage NodeStorage)
 			children[i] = child
 		}
 
-		return Branch{partialKey: key, children: children, value: NewFromEncoded(value)}, nil
+		return Branch{partialKey: key, children: children, value: NewValueFromEncoded(encoded.PartialKey, value)}, nil
+	default:
+		panic("unreachable")
+	}
+}
+
+type NodeToEncode interface {
+	isNodeToEncode()
+}
+
+type (
+	NewNodeToEncode struct {
+		partialKey []byte
+		value      []byte
+	}
+	TrieNodeToEncode struct {
+		child NodeHandle
+	}
+)
+
+func (NewNodeToEncode) isNodeToEncode()  {}
+func (TrieNodeToEncode) isNodeToEncode() {}
+
+type ChildReference interface {
+	isChildReference()
+	getNodeData() []byte
+}
+
+type (
+	HashChildReference struct {
+		hash common.Hash
+	}
+	InlineChildReference struct {
+		encodedNode []byte
+	}
+)
+
+func (HashChildReference) isChildReference() {}
+func (h HashChildReference) getNodeData() []byte {
+	return h.hash.ToBytes()
+}
+func (InlineChildReference) isChildReference() {}
+func (i InlineChildReference) getNodeData() []byte {
+	return i.encodedNode
+}
+
+type childFunc = func(node NodeToEncode, partialKey []byte, childIndex *byte) (ChildReference, error)
```

Comment:
could you define what a childFunc does?

Reply:
Sure, this is just an alias for a closure we execute when we are storing child nodes, you can see an example [here](https://github.com/ChainSafe/gossamer/blob/0eebe2d8cd2dc4c2befcca60285c4d6b297d5b26/pkg/trie/triedb/triedb.go#L807-L841)

Reply:
I see, can we rename it to `onChildStoreFn`, since this function stores functions, also do it need to be a closure, cannot it be an actual fn or routine that is called? I think that the need for a closure is when there are differences on how you implement something, eg. a predicate closure when might exists different conditions?

Reply:
Renamed! we use a closure to use free variables, for example [here](https://github.com/ChainSafe/gossamer/blob/0eebe2d8cd2dc4c2befcca60285c4d6b297d5b26/pkg/trie/triedb/triedb.go#L808) with the `prefixKey`.
I can create an issue to revisit it later if you want